### PR TITLE
SendInLoop Interprocedural Improvement

### LINKED ISSUE #174: `SendInLoop`: Interprocedural Improvement
**Summary**
Enhance the  #168 Send in Loop detector by introducing interprocedural analysis. This will allow the detector to identify send function calls made indirectly via other functions invoked within loops. 

**Context**
Currently, the Send in Loop detector only identifies send calls that happen directly inside loops. With interprocedural improvement, the detector will be able to track send calls even if they occur in other functions that are called from within a loop. This ensures more comprehensive detection of potential issues and is a key improvement planned for the next release.

This improvement will require implementing a #91 Callgraph and recursive analysis of function calls.

### PR DESCRIPTION
**Pull request by Esorat**
Closes #174 

- [x] I have updated `CHANGELOG.md`
- [x] I have added tests to demonstrate the contribution is correctly implemented
- [x] No test failures were reported when running `yarn test-all`
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: src/detectors/builtin/sendInLoop.ts [hunk 1]
```diff
@@ -76,13 +140,69 @@ export class SendInLoop extends ASTDetector {
...
+  private getFunctionsLeadingToSend(
+    callGraph: CallGraph,
+    functionsCallingSend: Set<CGNodeId>,
+  ): Set<CGNodeId> {
+    const functionsLeadingToSend = new Set<CGNodeId>(functionsCallingSend);
+
+    // Use a queue for BFS
+    const queue: CGNodeId[] = Array.from(functionsCallingSend);
+
+    while (queue.length > 0) {
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:186]: Why? We already have `areConnected` in the callgraph for this.

File: src/detectors/builtin/sendInLoop.ts [hunk 2]
```diff
@@ -103,4 +223,43 @@ export class SendInLoop extends ASTDetector {
       stmt.kind === "statement_foreach"
     );
   }
+
+  private getFunctionName(
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:227]: Duplicates logic from the callgraph?

File: src/detectors/builtin/sendInLoop.ts [hunk 3]
```diff
@@ -103,4 +223,43 @@ export class SendInLoop extends ASTDetector {
...
+    if (expr.kind === "static_call") {
+      return idText(expr.function);
+    } else if (expr.kind === "method_call") {
+      return idText(expr.method);
+    }
+    return undefined;
+  }
+
+  // Helper method to traverse expressions
+  private forEachExpression(
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:252]: Please remove GPT-generated code.

File: src/internals/ir/callGraph.ts [hunk 4]
```diff
@@ -13,7 +13,7 @@ import {
   AstStaticCall,
 } from "@tact-lang/compiler/dist/grammar/ast";
 
-type CGNodeId = number & { readonly brand: unique symbol };
+export type CGNodeId = number & { readonly brand: unique symbol };
 type CGEdgeId = number & { readonly brand: unique symbol };
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:17]: Should be public as well.

**Review by jubnzv — CHANGES_REQUESTED** [review]

File: src/internals/ir/callGraph.ts [hunk 5]
```diff
@@ -180,10 +176,24 @@ export class CallGraph {
...
+        );
+      } else {
+        this.logger.warn(`Caller function with AST ID ${func.id} not found.`);
+      }
+    }
+  }
+
   /**
    * Processes an expression, identifying static and method calls to add edges.
-   * @param expr The AST expression to process.
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:185]: Why?

File: src/internals/ir/callGraph.ts (hunk 4 above)
Comment (jubnzv) [reply src/internals/ir/callGraph.ts:17]: Why is it resolved?

File: src/detectors/builtin/sendInLoop.ts [hunk 6]
```diff
@@ -76,8 +116,46 @@ export class SendInLoop extends ASTDetector {
...
+      );
+
+      // Check function calls within the loop that lead to send
+      foldExpressions(
+        stmt,
+        (acc: MistiTactWarning[], expr: AstExpression) => {
+          if (expr.kind === "static_call" || expr.kind === "method_call") {
+            const calleeName = this.getCalleeName(expr);
+            if (calleeName) {
+              const calleeNodeId = callGraph.getNodeIdByName(calleeName);
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:129]: Will it work for methods?

File: src/detectors/builtin/sendInLoop.ts [hunk 7]
```diff
@@ -103,4 +181,13 @@ export class SendInLoop extends ASTDetector {
       stmt.kind === "statement_foreach"
     );
   }
+
+  private getCalleeName(expr: AstExpression): string | undefined {
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:185]: Why don't you use https://github.com/nowarp/misti/blob/e1fe0238cb7da1dcd616e8b75b81bc0e810ccd19/src/internals/ir/callGraph.ts#L168? I think we should generalize its functionality creating a function that gets the name from the function call.

**Review by jubnzv — CHANGES_REQUESTED** [review]
There is also a problem in the callgraph builder. We have name collisions for methods and free functions:
```
fun test() {}
contract C {
  fun test() { test() }
}
```
Generates the following CG:

изображение

Thus, we need to fix it first, and then use the updated `getFunctionName` function in this lint.

**Comment by jubnzv** [comment]
We also need to update `internals/ir/index.ts`:

```
export * from "./callgraph";
```

File: src/detectors/builtin/sendInLoop.ts [hunk 8]
```diff
@@ -43,29 +48,81 @@ export class SendInLoop extends ASTDetector {
...
+    const callGraph = new CallGraph(ctx).build(astStore);
+    const functionsCallingSend = new Set<CGNodeId>();
 
-    Array.from(cu.ast.getProgramEntries()).forEach((node) => {
-      forEachStatement(node, (stmt) => {
-        const warnings = this.analyzeStatement(stmt, processedLoopIds);
-        allWarnings.push(...warnings);
-      });
-    });
+    // Identify all functions that contain a send call
```
Comment (jubnzv) [inline src/detectors/builtin/sendInLoop.ts:56]: Let's do it once in `CallGraphBuilder` partially implementing #189. You could e.g. add a bitmask field to `Node` containing some information about it.

File: src/internals/ir/callGraph.ts [hunk 9]
```diff
@@ -247,3 +442,12 @@ export class CallGraph {
     }
   }
 }
+
+/**
+ * Helper function to check if an expression represents 'self'.
+ * @param expr The expression to check.
+ * @returns True if the expression is 'self'; otherwise, false.
+ */
+function isSelf(expr: AstExpression): boolean {
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:449]: https://github.com/nowarp/misti/blob/ae1f8d507cb421a1d943acb0d9c89e5f3420e473/src/internals/tact/util.ts#L72

File: src/internals/ir/callGraph.ts [hunk 10]
```diff
@@ -11,17 +11,32 @@ import {
...
-type CGEdgeId = number & { readonly brand: unique symbol };
+export type CGNodeId = number & { readonly brand: unique symbol };
+export type CGEdgeId = number & { readonly brand: unique symbol };
+
+/**
+ * Flag constants for CGNode.
+ */
+const FLAG_CALLS_SEND = 0b0001; // 1
+// You can define more flags as needed
+// const FLAG_ANOTHER_PROPERTY = 0b0010; // 2
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts]: Remove GPT-generated comments

File: src/internals/ir/callGraph.ts [hunk 11]
```diff
@@ -11,17 +11,32 @@ import {
...
+  AstContractDeclaration,
 } from "@tact-lang/compiler/dist/grammar/ast";
 
-type CGNodeId = number & { readonly brand: unique symbol };
-type CGEdgeId = number & { readonly brand: unique symbol };
+export type CGNodeId = number & { readonly brand: unique symbol };
+export type CGEdgeId = number & { readonly brand: unique symbol };
+
+/**
+ * Flag constants for CGNode.
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:23]: I would comment what this specific flag means

File: src/internals/ir/callGraph.ts [hunk 12]
```diff
@@ -52,64 +70,89 @@ class CGNode {
...
- * Nodes and edges are uniquely identified using indices generated by `IdxGenerator`.
- * This class provides methods to construct the graph from AST data, retrieve nodes and edges,
- * and check connectivity between nodes.
+ * The `CallGraph` class provides methods to build the graph from a Tact AST, retrieve nodes/edges,
+ * analyze connectivity, and add function calls dynamically.
  */
 export class CallGraph {
   private nodeMap: Map<CGNodeId, CGNode> = new Map();
-  private edgesMap: Map<CGEdgeId, CGEdge> = new Map();
+  private astIdToNodeId: Map<number, CGNodeId> = new Map();
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:92]:
```suggestion
  private astIdToNodeId: Map<AstNode["id"], CGNodeId> = new Map();
```

Correct?

File: src/internals/ir/callGraph.ts [hunk 13]
```diff
@@ -139,82 +182,234 @@ export class CallGraph {
...
+      }
+    }
+  }
+
+  /**
+   * Determines if the given expression is a 'send' call.
+   * @param expr The expression to check.
+   * @returns True if the expression is a 'send' call; otherwise, false.
+   */
+  private isSendCall(expr: AstExpression): boolean {
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:340]: I would move this to `src/internals/tact/util.ts` in order to make it available for other lints.

Comment (Esorat) [reply src/internals/ir/callGraph.ts:340]: move it

File: src/internals/ir/callGraph.ts (hunk 12 above)
Comment (Esorat) [reply src/internals/ir/callGraph.ts]: No, we cant remove edgesMap.  Part of logic.

File: src/internals/ir/callGraph.ts (hunk 11 above)
Comment (Esorat) [reply src/internals/ir/callGraph.ts:25]: Add

File: src/internals/ir/callGraph.ts (hunk 9 above)
Comment (Esorat) [reply src/internals/ir/callGraph.ts]: add export

File: src/internals/ir/callGraph.ts [hunk 14]
```diff
@@ -139,82 +185,217 @@ export class CallGraph {
...
+            declaration.kind === "receiver"
+          ) {
+            const func = declaration as
+              | AstFunctionDef
+              | AstContractInit
+              | AstReceiver;
+            const funcNodeId = this.astIdToNodeId.get(func.id);
+            if (funcNodeId !== undefined) {
+              let callsSend = false;
+              forEachExpression(func, (expr) => {
```
Comment (jubnzv) [inline src/internals/ir/callGraph.ts:302]: Use `findInExpressions`: https://github.com/nowarp/misti/blob/9ee55c0678bf4bc7f92e50b738e82ff0a2974ae5/src/internals/tact/iterators.ts#L212

Comment (Esorat) [reply src/internals/ir/callGraph.ts:301]: add

File: test/detectors/SendInLoop.expected.out [hunk 15]
```diff
@@ -1,62 +1,53 @@
...
-Help: Consider refactoring to avoid calling send functions inside loops
-See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop
-
-[MEDIUM] SendInLoop: Send function called inside a loop
-test/detectors/SendInLoop.tact:93:13:
-  92 |         while (i < 5) {
-> 93 |             self.reply(Msg{ a: i }.toCell());
-                   ^
-  94 |             i = i + 1;
+[MEDIUM] SendInLoop: Function "ComplexIndirectSend::deepNestSend" called inside a loop leads to a send function
```
Comment (jubnzv) [inline test/detectors/SendInLoop.expected.out:46]: It looks strange. Let's write that't a method or simply put the function name here.

Comment (Esorat) [reply test/detectors/SendInLoop.expected.out]: done

**Pull request merged**